### PR TITLE
New benchmark: Gesture semantics

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR adds a new benchmark that tests the performance of `GestureDetector`. It uses a new app, "Button Matrix", which contains 30 buttons on the screen, and repeatedly press buttons to update the screen. 

It also adds a new parameter `semanticsEnabled` to `benchmarkWidgets` in order to enable semantics, which defaults to `false` to avoid breaking existing benchmarks.

### Open issues
- Is the new app at its correct directory?

## Related Issues

- It blocks https://github.com/flutter/flutter/pull/33936 in order to test performance.

## Tests

N/A

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import 'package:flutter_test/flutter_test.dart';
+
+import '../common.dart';
+import './apps/button_matrix_app.dart' as button_matrix;
+
+const int _kNumWarmUpIters = 20;
+const int _kNumIters = 300;
+
+Future<void> main() async {
+  assert(false, "Don't run benchmarks in checked mode! Use 'flutter run --release'.");
```
Comment (dnfield) [inline dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart]: shouldn't this be `flutter run --profile`?

File: dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+
+    Future<void> iter() async {
+      // Press a button to update the screen
+      await tester.tapAt(const Offset(760.0, 30.0));
+      await tester.pump();
+    }
+
+    for (int i = 0; i < _kNumWarmUpIters; i += 1) {
```
Comment (dnfield) [inline dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart]: Maybe a comment to explain why we want to do warmups

File: dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart (hunk 1 above)
Comment (dnfield) [reply dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart]: This is also redundant with the check in `benchmarkWidgets`, or no?

Comment (dkwingsmt) [reply dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart]: I'm using the same boilerplate with other benchmarks, such as https://github.com/flutter/flutter/blob/master/dev/benchmarks/microbenchmarks/lib/stocks/animation_bench.dart

Comment (dnfield) [reply dev/benchmarks/microbenchmarks/lib/gestures/gesture_detector_bench.dart]: Ahh ok

**Review by dnfield — APPROVED** [review]
LGTM

**Comment by tvolkert** [comment]
For archaeologists, perf benchmark regressions triggered by this PR were filed at https://github.com/flutter/flutter/issues/35294

**Pull request merged**